### PR TITLE
Extend 'update-clouds' to update controller

### PR DESCRIPTION
**Pull request by Veebers**
## Description of change

'juju update-clouds' can update the cloud definitions on the controller
if the controller name is passed as an argument.

### CODE REVIEW

File: apiserver/facades/client/cloud/cloud.go [hunk 1]
```diff
@@ -28,6 +28,23 @@ import (
 
 var logger = loggo.GetLogger("juju.apiserver.cloud")
 
+// CloudV4 defines the methods on the cloud API facade, version 4.
+type CloudV4 interface {
```
Comment (wallyworld) [inline apiserver/facades/client/cloud/cloud.go]: You *could* embed CloudV3 here and just add the new methods

File: apiserver/facades/client/cloud/cloud.go [hunk 2]
```diff
@@ -896,6 +928,25 @@ func (api *CloudAPI) AddCloud(cloudArgs params.AddCloudArgs) error {
 	return nil
 }
 
+// UpdateCloud updates an existing cloud that the controller knows about.
+func (api *CloudAPI) UpdateCloud(cloudArgs params.UpdateCloudArgs) (params.ErrorResults, error) {
```
Comment (wallyworld) [inline apiserver/facades/client/cloud/cloud.go]: Needs a permission check. I wonder why Add etc doesn't have one. Ooops

File: apiserver/params/cloud.go [hunk 3]
```diff
@@ -28,6 +28,16 @@ type AddCloudArgs struct {
 	Name  string `json:"name"`
 }
 
+type CloudArgs struct {
```
Comment (wallyworld) [inline apiserver/params/cloud.go]: We can just reuse AddCloudArgs

File: state/cloud.go [hunk 4]
```diff
@@ -203,6 +237,23 @@ func (st *State) AddCloud(c cloud.Cloud, owner string) error {
 	return nil
 }
 
+// UpdateCloud updates an existing cloud with the given name and details.
+// Note that the Config is deliberately ignored - it's only
+// relevant when bootstrapping.
+func (st *State) UpdateCloud(c cloud.Cloud) error {
+	if err := validateCloud(c); err != nil {
+		return errors.Annotate(err, "invalid cloud")
```
Comment (wallyworld) [inline state/cloud.go]: if validateCloud() error is comprehensive, wrapping with "invalid cloud" doesn't really add much

File: apiserver/facades/client/cloud/cloud.go [hunk 5]
```diff
@@ -95,19 +118,28 @@ type CloudAPIV1 struct {
...
-func NewFacadeV3(context facade.Context) (*CloudAPI, error) {
+// NewFacadeV4 is used for API registration.
+func NewFacadeV4(context facade.Context) (*CloudAPI, error) {
 	st := NewStateBackend(context.State())
 	pool := NewModelPoolBackend(context.StatePool())
 	ctlrSt := NewStateBackend(pool.SystemState())
 	return NewCloudAPI(st, ctlrSt, pool, context.Auth(), state.CallContext(context.State()))
 }
 
+func NewFacadeV3(context facade.Context) (*CloudAPIV3, error) {
```
Comment (wallyworld) [inline apiserver/facades/client/cloud/cloud.go]: Comment

File: apiserver/facades/client/cloud/cloud.go [hunk 6]
```diff
@@ -889,13 +921,44 @@ func (api *CloudAPI) Credential(args params.Entities) (params.CloudCredentialRes
 
 // AddCloud adds a new cloud, different from the one managed by the controller.
 func (api *CloudAPI) AddCloud(cloudArgs params.AddCloudArgs) error {
-	err := api.backend.AddCloud(common.CloudFromParams(cloudArgs.Name, cloudArgs.Cloud), api.apiUser.Name())
+	isAdmin, err := api.authorizer.HasPermission(permission.SuperuserAccess, api.ctlrBackend.ControllerTag())
```
Comment (wallyworld) [inline apiserver/facades/client/cloud/cloud.go]: err will be NotFound if no permission has been recorded so you need something like this
(and for update etc too)

```
	isAdmin, err := api.authorizer.HasPermission(permission.SuperuserAccess, api.ctlrBackend.ControllerTag())
	if err != nil && !errors.IsNotFound(err) {
		return result, errors.Trace(err)
	} else if !isAdmin {
		return common.ServerError(common.ErrPerm)
	}

```

File: apiserver/params/cloud.go [hunk 7]
```diff
@@ -28,6 +28,11 @@ type AddCloudArgs struct {
 	Name  string `json:"name"`
 }
 
+// UpdateCloudArgs holds a cloud to be added with its name
```
Comment (wallyworld) [inline apiserver/params/cloud.go]: s/added/updated
Also end with "."

File: cmd/juju/cloud/add.go [hunk 8]
```diff
@@ -602,15 +580,68 @@ func queryCloudType(pollster *interact.Pollster) (string, error) {
...
+	}
+	personalClouds[newCloud.Name] = newCloud
+	return cloudMetadataStore.WritePersonalCloudMetadata(personalClouds)
+}
+
+type cloudFileReader struct {
+	cloudMetadataStore CloudMetadataStore
+}
+
+func (p cloudFileReader) ReadCloudFromFile(cloud, cloudFile string, ctxt *cmd.Context, verifyName bool) (*jujucloud.Cloud, error) {
```
Comment (wallyworld) [inline cmd/juju/cloud/add.go]: public methods need comment
does this need to be public?

File: cmd/juju/cloud/updatecloud.go [hunk 9]
```diff
@@ -0,0 +1,179 @@
+// Copyright 2018 Canonical Ltd.
```
Comment (wallyworld) [inline cmd/juju/cloud/updatecloud.go]: 2019

File: cmd/juju/cloud/updatecloud.go [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+
+var updateCloudDoc = `
+Update cloud information either locally or on the controller.
+
+Updating the local cache requires a <cloud name> and a yaml file containing the
+cloud details.
+
+To update a cloud on the controller you can provide just the <cloud name> which
+will use the cloud defined in the local cache or you can provide a cloud
+definition yaml file from which to retrieve the cloud details from.
```
Comment (wallyworld) [inline cmd/juju/cloud/updatecloud.go]: s/from//

File: cmd/juju/cloud/updatecloud.go [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+
+When <cloud definition file> is provided with <cloud name>, 
+Juju stores that definition its internal cache directly after
```
Comment (wallyworld) [inline cmd/juju/cloud/updatecloud.go]: in its

Also, we only want to do this if --controller is not specified

File: cmd/juju/cloud/updatecloud.go [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+    juju update-cloud mymaas -f path/to/maas.yaml
+    juju update-cloud mymaas -f path/to/maas.yaml --controller mycontroller
+    juju update-cloud mymaas --controller mycontroller
+
+See also:
+    add-cloud
+    list-clouds
+`
+
+type UpdateCloudAPI interface {
```
Comment (wallyworld) [inline cmd/juju/cloud/updatecloud.go]: does this need to be public?
needs a comment

File: cmd/juju/cloud/updatecloud.go [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+	}
+	if err := cmd.CheckEmpty(args[1:]); err != nil {
+		return err
+	}
+
+	// Condense arguments into an action,
+	c.commandAction = c.updateLocalCacheFromFile
+	if c.controllerName != "" {
+		if c.CloudFile != "" && c.Cloud != "" {
+			c.commandAction = c.updateControllerCacheFromFile
```
Comment (wallyworld) [inline cmd/juju/cloud/updatecloud.go]: updateControllerFromFile

It's not a controller cache

File: cmd/juju/cloud/updatecloud_test.go [hunk 14]
```diff
@@ -0,0 +1,161 @@
+// Copyright 2018 Canonical Ltd.
```
Comment (wallyworld) [inline cmd/juju/cloud/updatecloud_test.go]: 2019

File: cmd/juju/cloud/updatecloud_test.go [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+
+var _ = gc.Suite(&updateCloudSuite{})
+
+func (s *updateCloudSuite) SetUpTest(c *gc.C) {
+	s.FakeJujuXDGDataHomeSuite.SetUpTest(c)
+	s.api = &fakeUpdateCloudAPI{}
+	s.store = jujuclient.NewMemStore()
+}
+
+func (s *updateCloudSuite) TestBadArgs(c *gc.C) {
```
Comment (wallyworld) [inline cmd/juju/cloud/updatecloud_test.go]: I think there's some missing scenarios

File: cmd/juju/cloud/updatecloud_test.go [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+		Type:        "maas",
+		Description: "Metal As A Service",
+		AuthTypes:   jujucloud.AuthTypes{"oauth1"},
+		Endpoint:    "http://garagemaas",
+	})
+}
+
+type fakeUpdateCloudAPI struct {
+	jujutesting.Stub
+	updatedClouds []jujucloud.Cloud
```
Comment (wallyworld) [inline cmd/juju/cloud/updatecloud_test.go]: We set this but never check it.
We either need to record what was done and check that or use the stub to check the calls

File: cmd/juju/commands/main.go [hunk 17]
```diff
@@ -458,7 +458,8 @@ func registerCommands(r commandRegistry, ctx *cmd.Context) {
 	r.Register(setmeterstatus.New())
 
 	// Manage clouds and credentials
-	r.Register(cloud.NewUpdateCloudsCommand())
+	r.Register(cloud.NewUpdateCloudCommand(&cloudToCommandAdapter{}))
+	r.Register(cloud.NewUpdatePublicCloudsCommand())
```
Comment (wallyworld) [inline cmd/juju/commands/main.go]: I suspect main_test will need updating too

File: cmd/juju/cloud/updatecloud_test.go (hunk 16 above)
Comment (Veebers) [reply cmd/juju/cloud/updatecloud_test.go]: Good point, the raiseError was unused too (and we have NextErr() at any rate.). Calls are checked in the tests.

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**